aptos-move/move-examples/veiled_coin/sources/veiled_coin.move

```diff
@@ -0,0 +1,762 @@
+///
+/// ## Limitations
+///
+/// **WARNING:** This module is **experimental**! It is *NOT* production-ready. Specifically:
+///
+///  1. Deploying this module will likely lead to lost funds.
+///  2. This module has not been cryptographically-audited.
+///  3. The current implementation is vulnerable to _front-running attacks_ as described in the Zether paper [BAZB20].
+///  4. There is no integration with wallet software which, for veiled accounts, must maintain an additional ElGamal
+///    encryption keypair.
+///  5. There is no support for rotating the ElGamal encryption public key of a veiled account.
+///
+/// ## Veiled coin amounts as truncated `u32`'s
+///
+/// Veiled coin amounts must be specified as `u32`'s rather than `u64`'s as would be typical for normal coins in the
+/// Aptos framework. This is because coin amounts must be encrypted with an *efficient*, additively-homomorphic encryption
+/// scheme. Currently, our best candidate is ElGamal encryption in the exponent, which can only decrypt values around
+/// 32 bits or slightly larger.
+///
+/// Specifically, veiled coins are the middle 32 bits of the normal 64 bit coin values. In order to convert a `u32`
+/// veiled coin amount to a normal `u64` coin amount, we have to shift it left by 16 bits.
+///
+/// ```
+/// u64 normal coin amount format:
+/// [ left    || middle  || right ]
+/// [ 63 - 32 || 31 - 16 || 15 - 0]
+///
+/// u32 veiled coin amount format; we take the middle 32 bits from the u64 format above and store them in a u32:
+/// [ middle ]
+/// [ 31 - 0 ]
+/// ```
+///
+/// Recall that: A coin has a *decimal precision* $d$ (e.g., for `AptosCoin`, $d = 8$; see `initialize` in
+/// `aptos_coin.move`). This precision $d$ is used when displaying a `u64` amount, by dividing the amount by $10^d$.
+/// For example, if the precision $d = 2$, then a `u64` amount of 505 coins displays as 5.05 coins.
+///
+/// For veield coins, we can easily display a `u32` `Coin<T>` amount $v$ by:
+///  1. Casting $v$ as a u64 and shifting this left by 16 bits, obtaining a 64-bit $v'$
+///  2. Displaying $v'$ normally, by dividing it by $d$, which is the precision in `CoinInfo<T>`.
+///
+/// ## Implementation details
+///
+/// This module leverages a so-called "resource account," which helps us mint a `VeiledCoin<T>` from a
+/// normal `coin::Coin<T>` by transferring this latter coin into a `coin::CoinStore<T>` stored in the
+/// resource account.
+///
+/// Later on, when someone wants to convert their `VeiledCoin<T>` into a normal `coin::Coin<T>`,
+/// the resource account can be used to transfer out the normal from its coin store. Transfering out a coin like this
+/// requires a `signer` for the resource account, which the `veiled_coin` module can obtain via a `SignerCapability`.
+///
+/// ## TODOs
+///
+///  - We could have an `is_veiled` flag associated with the veiled balance, which we turn on only after a veiled to
```

Comment:
This feels like it would just confuse users and be redundant if we emphasize that `veil` transactions are public